obs-packaging/kata-containers-image/kata-containers-image.spec-template

```diff
@@ -32,6 +35,8 @@ install -m 0400 -p "${image}" ${ImageDir}/
 install -m 0400 -p "${initrd}"  ${ImageDir}/
 ln -s /usr/share/kata-containers/$(basename "${image}") ${ImageDir}/kata-containers.img
 ln -s /usr/share/kata-containers/$(basename "${initrd}")  ${ImageDir}/kata-containers-initrd.img
+chown -R root:kvm ${ImageDir}
+chmod -R g+rw ${ImageDir}
```

Comment:
afaik write access is not required, why you're adding it?

Reply:
@devimc Without the "w" permission, I get this error
```
Error: failed to launch qemu: exit status 1, error messages from qemu log: qemu-system-x86_64: -object memory-backend-file,id=mem0,mem-path=/usr/share/kata-containers/kata-containers-image_clearlinux_1.10.0-alpha1_agent_c87f497312.img,size=134217728: can't open backing store /usr/share/kata-containers/kata-containers-image_clearlinux_1.10.0-alpha1_agent_c87f497312.img for guest RAM: Permission denied: OCI runtime permission denied error
```
The permissions on the image are as : 
```
ls -la /usr/share/kata-containers/kata-containers-image_clearlinux_1.10.0-alpha1_agent_c87f497312.img
-rw-r--r-- 1 root root 134217728 Nov 20 22:36 /usr/share/kata-containers/kata-containers-image_clearlinux_1.10.0-alpha1_agent_c87f497312.img
```
with qemu being launched as non-root.

Reply:
@devimc Looks like the error comes from here : https://github.com/qemu/qemu/blob/master/exec.c#L1789

As you can see the file is being opened as RD_WR.
With root permissions this succeeds but fails when run as rootless, if the file itself does not have `rw` permissions.

Reply:
Could you show the complete qemu command being issued, as well as the configuration? I'm curious why RW access is needed for the image here. Thanks.

Reply:
@amshinde I'm worry about this change, we are giving rw access to anyone in the kvm group
@c3d  `.. -device nvdimm,id=nv0,memdev=mem0 -object memory-backend-file,id=mem0,mem-path=/usr/share/kata-containers/kata-containers-clearlinux-31770-osbuilder-36a9dc5-agent-88325f5.img,size=134217728 ..`  afaik currently in QEMU there is no option to open the backed file with ro, I sent a patch for ro filesystems but it needs more work [1]

[1] - https://patchew.org/QEMU/20190708211936.8037-1-julio.montes@intel.com/

Reply:
@devimc I share your concerns about having rw access, we should follow up on the patch you sent earlier. Looking at that thread, what is expected is having an explicit command line option, we should get some feedback on how the option should look like from qemu maintainers.

Reply:
cc @stefanha , since I see feedback from Stefan on the original patch.

Reply:
@amshinde Packaging question: why groupadd kvm instead of using a package dependency to ensure the group exists?  That would be cleaner, since you can specify the file ownership/permissions declaratively instead of relying on scripts that are executed during installation.

The semantics of the "kvm" group are that multiple semi-trusted users can utilize the KVM kernel module.  But these users do *not* trust each other.  Therefore group write permission is dangerous and cannot be allowed.  For example, I could replace your /usr/share/kata-containers/... image file to hijack your image so this is a security problem.

I'm also confused why the directory is made writable but the QEMU error message you posted is about the image file permissions.  These are two separate things.  Write permission on a directory means owners are allowed to create/delete/replace files but they are distinct from ownership permission on files in that directory.

Regarding the QEMU error message, if Kata requires writable pages but also needs the underlying file to be read-only, then it should be opened O_RDONLY but mmapped PROT_READ | PROT_WRITE with MAP_PRIVATE so that the host kernel does copy-on-write.  @devimc's patch is for true read-only operation, I believe, which is good if Kata doesn't require writable pages.

Reply:
> @amshinde I'm worry about this change, we are giving rw access to anyone in the kvm group

@amshinde Agreed, and also agree with @stefanha that this is probably not what you want.

Having seen the error message and where it comes from, I'm a bit further than "worry" and pretty much ready to "nack" this.

> @c3d `.. -device nvdimm,id=nv0,memdev=mem0 -object memory-backend-file,id=mem0,mem-path=/usr/share/kata-containers/kata-containers-clearlinux-31770-osbuilder-36a9dc5-agent-88325f5.img,size=134217728 ..` afaik currently in QEMU there is no option to open the backed file with ro, I sent a patch for ro filesystems but it needs more work [1]
> 
> [1] - https://patchew.org/QEMU/20190708211936.8037-1-julio.montes@intel.com/

My understanding is that images in /usr/share/kata-containers are intended to be read-only as far as qemu is concerned. As @stefanha pointed out, you could map them private and have a transient copy-on-write. Alternatively, you could make a container-private copy, e.g. under /var, and use that.

Reply:
@stefanha @c3d Thanks for your input. Marking the entire directory as rw was a quick dirty check to see if this passed the rootless test. I should have perhaps marked as wip, I intend to just add read permissions on that directory for `kvm`, while reconsidering the permissions on the image itself.
In case of Kata we do need a true read-only operation, we now mount the image as read-only, so @devimc's patch (with an extra condition to check for EACCESS error) should work for Kata. We do need to follow up on the patch upstream. @stefanha,  let us know if you have any suggestions on the qemu patch itself before we bring it up on the qemu mailing list.